### PR TITLE
Jdansariaws feature s3 cloudfront sap

### PR DESCRIPTION
**Pull request by JunaidAnsari80**
Issue #, if available:

This sample project demonstrates how to use AWS CloudFront and S3 to host a single-page application (SPA). The pattern creates a CloudFront distribution that serves the SPA content from an S3 bucket configured for static website hosting.:

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: s3-cloudfront-spa/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# S3 Hosted Website Served by a CloudFront Distribution restricted by Cloudfront Origin Access Control (OAC)
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:1]:
```suggestion
# Amazon S3 hosted website served by a CloudFront distribution restricted by CloudFront Origin Access Control (OAC)
```

File: s3-cloudfront-spa/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+This repo contains serverless patterns showing how to setup a S3 website hosting bucket that is served by a CloudFront distribution that also obfuscates the CloudFront Distribution domain via Cloudfront Origin Access Control (OAC).
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:3]: What do you mean by obfuscate the domain?

File: s3-cloudfront-spa/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+Important: this application uses various AWS services and there are costs associated with these services after the Free Tier usage - please see the [AWS Pricing page](https://aws.amazon.com/pricing/) for details. You are responsible for any AWS costs incurred. No warranty is implied in this example.
+
+## Requirements
+
+* AWS Account
+* AWS CLI already configured with Administrator permission
+* Terraform
+* Clone this repo!
+
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:18]: Please use the format from the template: https://github.com/aws-samples/serverless-patterns/tree/main/_pattern-model

File: s3-cloudfront-spa/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+1. Run terraform apply -destroy
+
+```
+git clone https://github.com/aws-samples/serverless-patterns/s3-cloudfront-spa
+```
+
+Each subdirectory contains additional installation and usage instructions. 
+
+----
+Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:38]:
```suggestion
Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: s3-cloudfront-spa/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+* Terraform
+* Clone this repo!
+
+## Deployment Instructions
+
+1. Go to s3-cloudfront-spa and copy template.tf to your project
+2. Setup terraform backend and define your state file S3 bucket by replacing your bucket name with "my-projects-terraform-state" in template.tf
+3. Run terraform init
+4. Run terraform plan
+5. Run terraform apply
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:25]: Please include the actual commands to run for point 2 and use proper code formatting

File: s3-cloudfront-spa/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+### Removing the resources
+
+1. Run terraform apply -destroy
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:29]: Is that correct?

File: s3-cloudfront-spa/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+* Clone this repo!
+
+## Deployment Instructions
+
+1. Go to s3-cloudfront-spa and copy template.tf to your project
+2. Setup terraform backend and define your state file S3 bucket by replacing your bucket name with "my-projects-terraform-state" in template.tf
+3. Run terraform init
+4. Run terraform plan
+5. Run terraform apply
+
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:26]: Please also include testing instructions to allow the user to verify that the pattern works as intended

File: s3-cloudfront-spa/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+### Removing the resources
+
+1. Run terraform apply -destroy
+
+```
+git clone https://github.com/aws-samples/serverless-patterns/s3-cloudfront-spa
+```
+
+Each subdirectory contains additional installation and usage instructions.
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:35]: What do you mean?

File: s3-cloudfront-spa/README.md
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:1]: please include tf in the parent folder name to indicate the framework used

File: s3-cloudfront-spa/example-pattern.json [hunk 9]
```diff
@@ -0,0 +1,74 @@
+
+{
+  "title": "AWS CloudFront and S3 for Single Page Application Hosting",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:3]:
```suggestion
  "title": "AWS CloudFront and Amazon S3 for Single Page Application Hosting",
```

File: s3-cloudfront-spa/example-pattern.json [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  "description": "This sample project demonstrates how to use AWS CloudFront and S3 to host a single-page application (SPA). The pattern creates a CloudFront distribution that serves the SPA content from an S3 bucket configured for static website hosting.",
+  "level": "200",
+  "framework": "terraform",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:6]:
```suggestion
  "framework": "Terraform",
```

File: s3-cloudfront-spa/example-pattern.json [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    "text": [
+      "terraform apply -destroy"
+    ]
+  },
+  "authors": [
+    {
+      "name": "Junaid Ansari",
+      "image": "https://media.licdn.com/dms/image/v2/C4E03AQEekYWyjMINfg/profile-displayphoto-shrink_200_200/profile-displayphoto-shrink_200_200/0/1659715539214?e=1728518400&v=beta&t=HVpwkCvJ_QdUYDW2YXbCabZDcWKWv3VpTBKmnCS5uVg",
+      "bio": "Cloud Application Architect at AWS.",
+      "linkedin": "https://www.linkedin.com/in/junaid-ansari-2734a912/"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:71]:
```suggestion
      "linkedin": "junaid-ansari-2734a912"
```

File: s3-cloudfront-spa/example-pattern.json [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      "Upload front end content. Copy Index.html from current repository for testing",
+      "Go to AWS Console",
+      "Go to AWS Cloudfront",
+      "Go to Distributions",
+      "Select your distribution domain and browse it. e.g ***********.cloudfront.net/Index.html"
+    ]
+  },
+  "cleanup": {
+    "text": [
+      "terraform apply -destroy"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:63]: Please double check and use <code> tags

File: s3-cloudfront-spa/example-pattern.json [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  "deploy": {
+    "text": [
+      "terraform init",
+      "terraform plan",
+      "terraform apply"
+    ]
+  },
+  "testing": {
+    "text": [
+      "Go to S3",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:53]: Refer to the README instead

File: s3-cloudfront-spa/example-pattern.json [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      "repoURL": "https://github.com/aws-samples/serverless-patterns/tree/main/s3-cloudfront-spa",
+      "templateURL": "serverless-patterns/s3-cloudfront-spa",
+      "projectFolder": "s3-cloudfront-spa",
+      "templateFile": "template.tf"
+    }
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Update terraform backend for s3 state file location"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:40]: This sections should include links to relevant content. Please refer to the template: https://github.com/aws-samples/serverless-patterns/blob/main/_pattern-model/example-pattern.json

File: s3-cloudfront-spa/example-pattern.json [hunk 15]
```diff
@@ -0,0 +1,74 @@
+
+{
+  "title": "AWS CloudFront and S3 for Single Page Application Hosting",
+  "description": "This sample project demonstrates how to use AWS CloudFront and S3 to host a single-page application (SPA). The pattern creates a CloudFront distribution that serves the SPA content from an S3 bucket configured for static website hosting.",
+  "level": "200",
+  "framework": "terraform",
+  "introBox": {
+    "headline": "How it works",
+    "text": [
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:9]: Please reword this to be no more than 4-6 full sentences

File: s3-cloudfront-spa/example-pattern.json [hunk 16]
```diff
@@ -0,0 +1,74 @@
+
+{
+  "title": "AWS CloudFront and S3 for Single Page Application Hosting",
+  "description": "This sample project demonstrates how to use AWS CloudFront and S3 to host a single-page application (SPA). The pattern creates a CloudFront distribution that serves the SPA content from an S3 bucket configured for static website hosting.",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:4]: Please reword to at most 150 characters

File: s3-cloudfront-spa/example-pattern.json [hunk 17]
```diff
@@ -0,0 +1,74 @@
+
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:1]: remove blank line

File: s3-cloudfront-spa/template.tf [hunk 18]
```diff
@@ -0,0 +1,145 @@
+terraform {
+    backend "s3" {
+      key = "terraform.tfstate"
+      region = "eu-west-2"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/template.tf:4]: Could this be made configurable easily? Or at least call it out in the deployment instructions

File: s3-cloudfront-spa/template.tf [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+      bucket = "my-projects-terraform-state"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/template.tf:5]: As bucket names must be unique, this will lead to problems.

File: s3-cloudfront-spa/template.tf [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+  custom_error_response {
+    error_code         = 404
+    response_code      = 200
+    response_page_path = "/index.html"
+  }
+
+  custom_error_response {
+    error_code         = 403
+    response_code      = 200
+    response_page_path = "/index.html"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/template.tf:78]: What is the purpose of this?

File: s3-cloudfront-spa/README.md (hunk 8 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md:35]: Removed

File: s3-cloudfront-spa/example-pattern.json (hunk 12 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:63]: https://developer.hashicorp.com/terraform/cli/commands/destroy

For simplicity I am going update the command to terrform destroy

File: s3-cloudfront-spa/example-pattern.json (hunk 13 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:53]: updated

File: s3-cloudfront-spa/template.tf (hunk 18 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/template.tf:4]: Yes i will make it configurable by adding variables for bankend key, region and bucket

File: s3-cloudfront-spa/example-pattern.json (hunk 17 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:1]: removed

File: s3-cloudfront-spa/example-pattern.json (hunk 16 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:4]: Updated

File: s3-cloudfront-spa/template.tf (hunk 18 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/template.tf:4]: I will remove the terraform remote setup

File: s3-cloudfront-spa/template.tf (hunk 19 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/template.tf:5]: Remove the Terraform remote setup. It is not mandatory for the pattern.

File: s3-cloudfront-spa/example-pattern.json (hunk 14 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:39]: implemented

File: s3-cloudfront-spa/example-pattern.json (hunk 15 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:8]: Updated the json and reduced to 6 steps

File: s3-cloudfront-spa/README.md (hunk 7 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md:47]: Added the testing section

File: s3-cloudfront-spa/README.md (hunk 5 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md]: updated the commands with formatting.

File: s3-cloudfront-spa/README.md (hunk 3 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md]: updated and applied the format

File: s3-cloudfront-spa/README.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
+# Amazon S3 hosted website served by a CloudFront distribution restricted by CloudFront Origin Access Control (OAC)
+This project serves as a practical example or demonstration of how to leverage two powerful services from Amazon Web Services (AWS) - Amazon CloudFront and Amazon Simple Storage Service (S3) - to host and deliver a static website efficiently and securely.
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:2]:
```suggestion
This project serves as a demonstration of how to leverage Amazon CloudFront and Amazon Simple Storage Service (S3) to host and deliver a static website efficiently and securely.
```

File: s3-cloudfront-spa/README.md [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+- To learn more about submitting a pattern, read the [publishing guidelines page](https://github.com/aws-samples/serverless-patterns/blob/main/PUBLISHING.md).
+
+Important: this application uses various AWS services and there are costs associated with these services after the Free Tier usage - please see the [AWS Pricing page](https://aws.amazon.com/pricing/) for details. You are responsible for any AWS costs incurred. No warranty is implied in this example.
+
+## Requirements
+
+* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
+* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
+* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+* [Terraform Install](https://www.terraform.io/)
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:20]:
```suggestion
* [Terraform Installed](https://www.terraform.io/)
```

File: s3-cloudfront-spa/README.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+6. Copy your front end assests
+
+## Testing
+    "Go to your hosting Amazon S3 bucket ",
+    "Upload your front end content or Copy index.html from "s3-cloudfront-spa" pattern "src" folder for testing",
+    "Go to AWS Console",
+    "Go to AWS Cloudfront",
+    "Go to Distributions",
+    "Select your distribution domain and browse it. e.g ***********.cloudfront.net/index.html"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/README.md:54]: Why is this section in quotes?

File: s3-cloudfront-spa/README.md
Comment (bfreiberg) [reply s3-cloudfront-spa/README.md:1]: Sorry, I might have been not clear enough. The parent folder name (`s3-cloudfront-spa`) should contain the framework abbreviation: `s3-cloudfront-spa-tf`

File: s3-cloudfront-spa/example-pattern.json [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+    ]
+  },
+  "testing": {
+    "text": [
+      "Refer to README"
+    ]
+  },
+  "cleanup": {
+    "text": [
+      "terraform destroy"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:52]:
```suggestion
      "<code>terraform destroy</code>"
```

File: s3-cloudfront-spa/example-pattern.json [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+      "templateURL": "serverless-patterns/s3-cloudfront-spa",
+      "projectFolder": "s3-cloudfront-spa",
+      "templateFile": "template.tf"
+    }
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Amazon S3 bucket to host the website",
+        "link": "https://aws.amazon.com/s3/"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:30]: Please use a more specific link

File: s3-cloudfront-spa/example-pattern.json [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+      },
+      {
+        "text": "Amazon CloudFront uses Amazon S3 as the origin server to host and serve static websites.",
+        "link": "https://aws.amazon.com/cloudfront/"
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:34]: Please use a more specific link

File: s3-cloudfront-spa/example-pattern.json [hunk 27]
```diff
@@ -0,0 +1,63 @@
+{
+  "title": "AWS CloudFront and Amazon S3 for Single Page Application Hosting",
+  "description": "This project shows how to host a Single-Page Application (SPA) using AWS CloudFront & S3. CloudFront serves the SPA content from an S3 bucket configured for static hosting.",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:3]: Please reword to stay below 150 characters

File: s3-cloudfront-spa/example-pattern.json [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+  "level": "200",
+  "framework": "Terraform",
```
Comment (bfreiberg) [inline s3-cloudfront-spa/example-pattern.json:5]: Please also add the language as a top level attribute

**Review by bfreiberg — CHANGES_REQUESTED** [review]

File: s3-cloudfront-spa/README.md (hunk 21 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md:2]: Implemented

File: s3-cloudfront-spa/README.md (hunk 23 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md:54]: Removed the quotes.

File: s3-cloudfront-spa/README.md (hunk 22 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/README.md:20]: updated as suggested

File: s3-cloudfront-spa/example-pattern.json (hunk 24 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:52]: updated the text with <code> tags for cleanup as well for the deploy

File: s3-cloudfront-spa/example-pattern.json (hunk 25 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:30]: I have updated the link to amazon s3 static hosting

File: s3-cloudfront-spa/example-pattern.json (hunk 26 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:34]: link updated

File: s3-cloudfront-spa/example-pattern.json (hunk 27 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:3]: updated to 149 characters

File: s3-cloudfront-spa/example-pattern.json (hunk 28 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa/example-pattern.json:5]: language added.

File: s3-cloudfront-spa-tf/README.md [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+
+6. Copy your front end assests
+
+## Testing
+1.  Go to your hosting Amazon S3 bucket.
+2.  Upload your front end content or Copy index.html from "s3-cloudfront-spa" pattern "src" folder for testing.
+3.  Go to AWS Console.
+4.  Go to AWS Cloudfront.
+5.  Go to Distributions.
+5.  Select your distribution domain and browse it. e.g ***********.cloudfront.net/index.html.
```
Comment (bfreiberg) [inline s3-cloudfront-spa-tf/README.md:54]: I'm sorry but this section still needs some work. For example, step 3 seems unnecessary as the user is already in the AWS console as they uploaded a file to S3.

File: s3-cloudfront-spa-tf/README.md [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    Run terraform apply
+    ```
+    During the prompts:
+    * Enter yes
+
+6. Copy your front end assests
+
+## Testing
+1.  Go to your hosting Amazon S3 bucket.
+2.  Upload your front end content or Copy index.html from "s3-cloudfront-spa" pattern "src" folder for testing.
```
Comment (bfreiberg) [inline s3-cloudfront-spa-tf/README.md:50]: no need to refer to the parent folder name. And please rename the `src` folder to `test` and refer to it here correctly

File: s3-cloudfront-spa-tf/example-pattern.json [hunk 31]
```diff
@@ -0,0 +1,64 @@
+{
+  "title": "AWS CloudFront and Amazon S3 for Single Page Application Hosting",
+  "description": "This project demonstrates hosting a Single-Page App using AWS CloudFront & S3, where CloudFront serves the SPA from an S3 bucket for static hosting.",
+  "language": "HCL",
```
Comment (bfreiberg) [inline s3-cloudfront-spa-tf/example-pattern.json:4]:
```suggestion
  "language": "Terraform",
```

Comment (bfreiberg) [reply s3-cloudfront-spa-tf/example-pattern.json:4]: I know this seems incorrect but this is just how it is represented internally

**Review by bfreiberg — CHANGES_REQUESTED** [review]

File: s3-cloudfront-spa-tf/example-pattern.json (hunk 31 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa-tf/example-pattern.json:4]: I have updated to Terraform

File: s3-cloudfront-spa-tf/README.md (hunk 30 above)
Comment (JunaidAnsari80) [reply s3-cloudfront-spa-tf/README.md:50]: src folder updated to test and instructions are updated.

**Review by bfreiberg — APPROVED** [review]

**Comment by bfreiberg** [comment]
Looks good, thank you for your contribution. The PR will be merged to Serverlessland.com soon

**Comment by JunaidAnsari80** [comment]
Thanks @bfreiberg for your support.
